shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-distsql/shardingsphere-db-discovery-distsql-handler/src/main/java/org/apache/shardingsphere/dbdiscovery/distsql/handler/update/CreateDatabaseDiscoveryRuleStatementUpdater.java

```diff
@@ -89,12 +90,22 @@ private void checkToBeCreatedResources(final String schemaName,
         }
     }
     
-    private void checkToBeCreatedDiscoverTypes(final CreateDatabaseDiscoveryRuleStatement sqlStatement) throws InvalidAlgorithmConfigurationException {
-        Collection<String> notExistedDiscoveryTypes = sqlStatement.getRules().stream().map(DatabaseDiscoveryRuleSegment::getDiscoveryTypeName).distinct()
+    private void checkDiscoverTypeAndHeartbeat(final CreateDatabaseDiscoveryRuleStatement sqlStatement, final DatabaseDiscoveryRuleConfiguration currentRuleConfig) throws DistSQLException {
+        Map<String, List<AbstractDatabaseDiscoverySegment>> segmentMap = sqlStatement.getRules().stream().collect(Collectors.groupingBy(each -> each.getClass().getSimpleName()));
+        Collection<String> invalidInput = segmentMap.getOrDefault(DatabaseDiscoveryDefinitionSegment.class.getSimpleName(), Collections.emptyList()).stream()
+                .map(each -> ((DatabaseDiscoveryDefinitionSegment) each).getDiscoveryType().getName()).distinct()
                 .filter(each -> !TypedSPIRegistry.findRegisteredService(DatabaseDiscoveryType.class, each, new Properties()).isPresent()).collect(Collectors.toList());
-        if (!notExistedDiscoveryTypes.isEmpty()) {
-            throw new InvalidAlgorithmConfigurationException("database discover", notExistedDiscoveryTypes);
-        }
+        DistSQLException.predictionThrow(invalidInput.isEmpty(), new InvalidAlgorithmConfigurationException("database discovery", invalidInput));
+        segmentMap.getOrDefault(DatabaseDiscoveryConstructionSegment.class.getSimpleName(), Collections.emptyList()).stream().map(each -> (DatabaseDiscoveryConstructionSegment) each)
+                .forEach(each -> {
+                    if (null == currentRuleConfig || !currentRuleConfig.getDiscoveryTypes().containsKey(each.getDiscoveryTypeName())) {
+                        invalidInput.add(each.getDiscoveryTypeName());
+                    }
+                    if (null == currentRuleConfig || !currentRuleConfig.getDiscoveryHeartbeats().containsKey(each.getDiscoveryHeartbeatName())) {
+                        invalidInput.add(each.getDiscoveryHeartbeatName());
+                    }
+                });
+        DistSQLException.predictionThrow(invalidInput.isEmpty(), new RequiredAlgorithmMissedException("database discovery", invalidInput));
```

Comment:
The string "database discovery" has been used many times. Consider declaring it as a class field?